### PR TITLE
Add valgrind memory checks to Azure pipeline.

### PR DESCRIPTION
**Pull request by hodgestar**
Continuation of #78 from @thedrow.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,3 +18,6 @@ cppcheck: cppcheck-build-dir
 infer:
 	python3 setup.py build_ext -if -U NDEBUG | compiledb
 	@infer --fail-on-bug --compilation-database compile_commands.json
+
+valgrind:
+	PYTHONMALLOC=malloc valgrind --suppressions=hpy/tools/valgrind/python.supp --suppressions=hpy/tools/valgrind/hpy.supp --leak-check=full --show-leak-kinds=definite,indirect --log-file=/tmp/valgrind-output python -m pytest --valgrind --valgrind-log=/tmp/valgrind-output test/
```
Comment (hodgestar) [inline Makefile:23]: @thedrow Should we add `--gen-suppressions=all` here? It makes it convenient to write new suppressions or to compare with existing ones, but not sure what's considered best practice.

**Comment by hodgestar** [comment]
@fangerer @rlamy @antocuni -- comments on the leak fixes would be appreciated.

File: hpy/devel/src/runtime/ctx_module.c [hunk 2]
```diff
@@ -29,8 +29,10 @@ ctx_Module_Create(HPyContext ctx, HPyModuleDef *hpydef)
     def->m_doc = hpydef->m_doc;
     def->m_size = hpydef->m_size;
     def->m_methods = create_method_defs(hpydef->defines, hpydef->legacy_methods);
-    if (def->m_methods == NULL)
+    if (def->m_methods == NULL) {
+        PyMem_Free(def);
```
Comment (antocuni) [inline hpy/devel/src/runtime/ctx_module.c:33]: LGTM

File: Makefile (hunk 1 above)
Comment (thedrow) [reply Makefile:23]: We could be it is very verbose.
Instead, when we encounter a memory leak and verify it is a false positive, we can run the specific test that triggers it and generate a suppression for that.

The suppressions I've generated so far appears to be either genuine leaks from Python (one of them appears to be fixed in Python 3.9) or genuine memory leaks in HPy that we're aware of and won't fix for now.

File: hpy/devel/src/runtime/ctx_type.c [hunk 3]
```diff
@@ -405,12 +411,15 @@ ctx_Type_FromSpec(HPyContext ctx, HPyType_Spec *hpyspec,
     }
     PyObject *bases = build_bases_from_params(params);
     if (PyErr_Occurred()) {
+        PyMem_Free(spec->slots);
+        PyMem_Free(spec);
         return HPy_NULL;
     }
     PyObject *result = PyType_FromSpecWithBases(spec, bases);
     /* note that we do NOT free the memory which was allocated by
```
Comment (thedrow) [inline hpy/devel/src/runtime/ctx_type.c:419]: For example this is an actual leak I've suppressed.

File: hpy/devel/src/runtime/ctx_type.c [hunk 4]
```diff
@@ -351,6 +351,7 @@ static PyObject *build_bases_from_params(HPyType_SpecParam *params)
     if (params == NULL)
         return NULL;
 
+    PyObject *tup;
```
Comment (antocuni) [inline hpy/devel/src/runtime/ctx_type.c:354]: The whole commit LGTM

File: hpy/universal/src/hpymodule.c [hunk 5]
```diff
@@ -113,7 +113,7 @@ static PyObject *load_from_spec(PyObject *self, PyObject *spec)
 
     HPy mod = ((InitFuncPtr)initfn)(&global_ctx);
     if (HPy_IsNull(mod))
-        return NULL;
+        goto error;
```
Comment (antocuni) [inline hpy/universal/src/hpymodule.c:116]: LGTM

File: test/test_hpydict.py [hunk 6]
```diff
@@ -60,6 +60,7 @@ def test_GetItem(self):
                 if (HPy_IsNull(key))
                     return HPy_NULL;
                 HPy val = HPyDict_GetItem(ctx, arg, key);
+                HPy_Close(ctx, key);
```
Comment (antocuni) [inline test/test_hpydict.py:63]: LGTM

**Review by antocuni — APPROVED** [review]

**Comment by antocuni** [comment]
All your commits seem to fix real leaks: I'd say this is a good sign that valgrind is actually useful, so am I +1 to merge this.

**Comment by hodgestar** [comment]
Thanks everyone, and especially @thedrow for suggesting this. Let me merge this.

**Pull request merged**